### DISCUSSION TITLE
Telnet from within a Switch

### DISCUSSION THREAD

**Discussion by remchuk** [napalm-automation/napalm#2097] [Q&A]
Hey there everyone!

I have a bunch of Catalyst 8300-2N2S-4T2X with a `16 ports Async Serial NIM`.

Is there a way for me using napalm to telnet from within the catalyst ?

to run `telnet <ip of the catayst> <line number>` and then push username and password?

**Comment by AdamRickards** [comment]
Is this still an open concern for you @remchuk ?

I was just browsing this discussion forum directed from the napalm slack, and it seems like an interesting problem.

I have not yet tested it but I was just thinking about a way to implement this that solves a problem I was looking to solve for myself (in my community driver) but also that might help you as well.

I put together some thoughts into Claude and iterated a few times to end up here, if you still have interest i'll test it on my own hardware (different target type, use case, but should work the same) and release it so you can give it a shot.

# napalm-telnet-hop

A NAPALM community driver for telnet hop-through sessions.

## What it does

SSH into a jump host, telnet hop to a target device (via terminal server reverse telnet, direct telnet, or any custom connect command), run commands, return output, exit cleanly.

The hop/exit mechanics are YAML-defined, making the driver platform-agnostic on both the jump host and target side.

## Use cases

- **Terminal server**: Catalyst 8300 with async serial NIM, Cisco 2900 with HWIC-16A, or any device acting as a console/terminal server. Reverse telnet through a serial line to reach the connected device.
- **Direct telnet**: SSH into any device, telnet from its CLI to a target that only speaks telnet.
- **Custom hop**: Any platform with a CLI command to reach another device — `connect serial`, `session`, or whatever your platform uses. Define it in YAML.

## How it works

1. Opens SSH to jump host via Netmiko
2. Sends hop command (formatted from YAML template + runtime args)
3. Walks through hop_sequence (if defined) — expect/respond steps
4. Authenticates to target device (auto-detects username/password prompts)
5. Sends each command with timing-based reads (no cmd_verify)
6. Walks through exit_sequence to return to jump host
7. Verifies jump host prompt is restored
8. Returns dict of command outputs

**Reply by remchuk** [reply]
Hey there Adam!
In the end what worked for me was to implement an SSH rather then telnet from within the switch.
It is still messy but at least it works.

I exec the commands in an order:
ssh -p 2010 <ip of the switch>
\n
password
\n
get the host name.
close session,


meaning that I have to exec the "enter" as I would do in a real telnet or ssh session made from within the switch.
